### DISCUSSION TITLE
client-side transition does not take into account that the target page has callback:true

### DISCUSSION THREAD

**Discussion by diegopelusa** [vercel/next.js#24265] [Ideas]
### What version of Next.js are you using?

10.1.3

### What version of Node.js are you using?

12.19.0

### What browser are you using?

Firefox, Chrome

### What operating system are you using?

Windows

### How are you deploying your application?

npm run build && npm run start (local machine)

### Describe the Bug

When doing a client-side transition to a static page with dynamic routing and fallback:true for the first time, the getStaticProps function is being executed,
```
export async function getStaticProps({ params }) {
...
}
```
 but the fallback version of the page is not being executed.
```
 if (router.isFallback) {
    return <div>Loading...</div>
  }
```
It seems that that in client-side transition the fallback of the target page is not being considered.
That makes, that the behaviour of the static page with dynamic routing behaves differently the first time, depending if is called from the browser (externally) or from a page via Link/router.push().

### Expected Behavior

I would expect that the case where fallback:true is taken into consideration in client-side transition.

A way to avoid this issue would be to call:
`window.location.assign(<base_url> + router.locale + <page_url>);`

But it would be nicer that the same call could be done with Link or router.push().

Maybe adding a new parameter for Link/router.push(), like pageFallback, where if:
- pageFallback = {true} -> Take into consideration the value of fallback of the goal page.
- pageFallback = {false} -> Do client-side transition as normal.

### To Reproduce

By calling in this way, when the target page is being first called, the fallback:true case is not executed:
```
onClick={(event) => {
                      router.push("/recipe/bla-02090");
                    }}
```

By calling in this way, when the target page is being first called, the fallback:true case is executed:
```
onClick={(event) => {
                      window.location.assign(
                        "http://localhost:3000/" +
                          router.locale +
                          "/recipe/bla-02090"
                      );
                    }}
```

**Comment by diegopelusa** [comment]
Hi @timneutkens !
You closed this issue and moved to "ideas".
Is it because it does not make sense to have a client-side transition with fallback:true?
Or just because it would be a new feature, that has to be discussed?

Do you see the point that the page being called externally or from an internal page will load differently for the first time?

Thanks for any explanation!
Cheers
Diego

**Comment by miguderp** [comment]
Unless if I'm mistaken the prefetch functionality of Next.js pre-builds your page as soon as your link enters your viewport, hence `router.isFallback` not being taken into account because the page already exists by the time you're about to click on it.

To verify this, maybe check your `.next` folder (on a freshly built project) and search if your dynamic page you're about to click is already there or not (not sure about the pathing but maybe under `.next/server/pages/etc/<<<your dynamic page slug>>>`)?

**Reply by diegopelusa** [reply]
I don´t think that prefetch has the same effect as static+fallback.
I don´t want to prefetch all static files which I want to navigate via a client-side transition (link or router.push). It could be thousands of pages to be prefetch!

The situation is the following:
1 - If I call a static page with dynamic routing directly at the navigation bar, then the NextJs either: 
1a- the static page was already created and then it is delivered inmediately
1b- the static page was not created, then NextJs checks the fallback parameter:
1b1- fallback: false -> NextJs delivers the 404 page
1b2- fallback: true -> NextJs creates the static page and in the meantime it delivers first code within router.isFallback
1b3- fallback: blocking -> NextJs creates the static page and nothing is delivered until the page is completely created

2- If I call a static page with dynamic routing from another page with a client-side transition (link or router.push), then the NextJs either:
2a- the static page was already created and then it is delivered immediately
2b- the static page was not already created and then in any case NextJs creates the page and nothing is delivered until the page is completely created. This corresponds o the case (1b3 fallback: blocking)
  
My question is:
Would it not be possible to NextJs to check the fallback parameter in client-side transition, as it does when the page is called "externally"?
In my opinion, the only thing that it would have to do is to replicate the same logic as when it is called externally, i.e.:
- fallback:false -> 404 page
- fallback:true -> while creating the page NextJs delivers the code within router.isFallback
- fallback:blocking -> (as it is working now) creates the page and delivers it when done.

**Reply by miguderp** [reply]
If you don't want to have the prefetch feature, you can `<Link prefetch={false} />` allowing you I believe to achieve your goal (or similar)
